File: src/bot/cogs/user_commands.py
"""User-related commands cog."""

import discord
from discord.ext import commands

from models import get_db
from services.database import UserService
from utils.logger import get_logger

logger = get_logger(__name__)


class UserCommands(commands.Cog):
    """Commands for user tracking and gamification."""

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @commands.command(name="profile", aliases=["stats", "me"])
    async def profile(self, ctx: commands.Context):
        """
        View your sustainability profile.

        Usage: !profile
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                user = UserService.get_or_create(db, ctx.author.id, ctx.author.name)

                embed = discord.Embed(
                    title=f"👤 {user.username}'s Profile",
                    color=discord.Color.blue()
                )

                embed.add_field(
                    name="🏆 Rank",
                    value=user.get_rank(),
                    inline=True
                )
                embed.add_field(
                    name="📊 Level",
                    value=str(user.level),
                    inline=True
                )
                embed.add_field(
                    name="⭐ Points",
                    value=str(user.sustainability_points),
                    inline=True
                )
                embed.add_field(
                    name="📝 Queries",
                    value=str(user.queries_count),
                    inline=True
                )

                # Progress to next level
                points_to_next = ((user.level) * 100) - user.sustainability_points
                if points_to_next > 0:
                    embed.add_field(
                        name="📈 Progress",
                        value=f"{points_to_next} points to level {user.level + 1}",
                        inline=True
                    )

                # Preferences
                if user.preferred_materials:
                    embed.add_field(
                        name="Preferred Materials",
                        value=user.preferred_materials,
                        inline=False
                    )

                if user.sustainability_priority:
                    embed.add_field(
                        name="Priority",
                        value=user.sustainability_priority.replace("_", " ").title(),
                        inline=True
                    )

                if user.budget_range:
                    embed.add_field(
                        name="Budget Range",
                        value=user.budget_range,
                        inline=True
                    )

                embed.set_thumbnail(url=ctx.author.display_avatar.url)
                embed.set_footer(text="Keep learning about sustainability to earn more points!")

                await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in profile command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred while fetching your profile.")
            finally:
                db.close()

    @commands.command(name="leaderboard", aliases=["lb", "top"])
    async def leaderboard(self, ctx: commands.Context):
        """
        View the sustainability leaderboard.

        Usage: !leaderboard
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                top_users = UserService.get_leaderboard(db, limit=10)

                if not top_users:
                    await ctx.send("❌ No users found on leaderboard.")
                    return

                embed = discord.Embed(
                    title="🏆 Sustainability Champions Leaderboard",
                    description="Top users making a difference:",
                    color=discord.Color.gold()
                )

                medals = ["🥇", "🥈", "🥉"]

                for i, user in enumerate(top_users, 1):
                    medal = medals[i-1] if i <= 3 else f"**{i}.**"
                    embed.add_field(
                        name=f"{medal} {user.username}",
                        value=f"Level {user.level} • {user.sustainability_points} points • {user.get_rank()}",
                        inline=False
                    )

                # Show current user's rank if not in top 10
                current_user = UserService.get_by_discord_id(db, ctx.author.id)
                if current_user and current_user not in top_users:
                    embed.add_field(
                        name=f"\nYour Stats",
                        value=f"Level {current_user.level} • {current_user.sustainability_points} points",
                        inline=False
                    )

                embed.set_footer(text="Use !profile to view your complete profile")

                await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in leaderboard command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred while fetching leaderboard.")
            finally:
                db.close()

    @commands.command(name="setpreference", aliases=["pref"])
    async def set_preference(
        self, ctx: commands.Context, preference_type: str, *, value: str
    ):
        """
        Set your sustainability preferences.

        Usage: !setpreference <type> <value>
        Types: materials, budget, priority
        Example: !setpreference materials organic cotton, linen
        """
        db = next(get_db())
        try:
            user = UserService.get_or_create(db, ctx.author.id, ctx.author.name)

            preference_type = preference_type.lower()

            if preference_type in ["materials", "material"]:
                user.preferred_materials = value
                await ctx.send(f"✅ Preferred materials set to: {value}")
            elif preference_type == "budget":
                if value not in ["$", "$$", "$$$", "$$$$"]:
                    await ctx.send("❌ Budget must be: $, $$, $$$, or $$$$")
                    return
                user.budget_range = value
                await ctx.send(f"✅ Budget range set to: {value}")
            elif preference_type == "priority":
                valid_priorities = ["environmental", "social", "animal_welfare", "all"]
                if value.lower() not in valid_priorities:
                    await ctx.send(f"❌ Priority must be one of: {', '.join(valid_priorities)}")
                    return
                user.sustainability_priority = value.lower()
                await ctx.send(f"✅ Sustainability priority set to: {value}")
            else:
                await ctx.send(
                    "❌ Invalid preference type. Use: materials, budget, or priority"
                )
                return

            UserService.update(db, user)

        except Exception as e:
            logger.error(f"Error in setpreference command: {e}", exc_info=True)
            await ctx.send("❌ An error occurred while setting preference.")
        finally:
            db.close()


async def setup(bot: commands.Bot):
    """Set up the cog."""
    await bot.add_cog(UserCommands(bot))

File: src/commands/user_commands.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
//! User cog: `profile`, `leaderboard`, `setpreference`.

use poise::serenity_prelude as serenity;

use super::{colour, say, send_embed};
use crate::services::UserService;
use crate::{typing, Context, Error};

fn title_case(s: &str) -> String {
    s.split('_')
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                Some(f) => f.to_uppercase().chain(c).collect::<String>(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// View your sustainability profile.
#[poise::command(prefix_command, aliases("stats", "me"))]
pub async fn profile(ctx: Context<'_>) -> Result<(), Error> {
    typing(&ctx).await;
    let db = &ctx.data().db;
    let user =
        UserService::get_or_create(db, ctx.author().id.get() as i64, &ctx.author().name).await?;

    let username = user.username.clone().unwrap_or_default();
    let mut embed = serenity::CreateEmbed::new()
        .title(format!("👤 {}'s Profile", username))
        .colour(colour::blue())
        .field("🏆 Rank", user.get_rank(), true)
        .field("📊 Level", user.level.to_string(), true)
        .field("⭐ Points", user.sustainability_points.to_string(), true)
        .field("📝 Queries", user.queries_count.to_string(), true);

    let points_to_next = user.level * 100 - user.sustainability_points;
    if points_to_next > 0 {
        embed = embed.field(
            "📈 Progress",
            format!(
                "{} points to level {}",
                points_to_next,
                user.level + 1
            ),
            true,
        );
    }
    if let Some(pm) = &user.preferred_materials {
        embed = embed.field("Preferred Materials", pm, false);
    }
    if let Some(p) = &user.sustainability_priority {
        embed = embed.field("Priority", title_case(p), true);
    }
    if let Some(b) = &user.budget_range {
        embed = embed.field("Budget Range", b, true);
    }
    embed = embed
        .thumbnail(ctx.author().face())
        .footer(serenity::CreateEmbedFooter::new(
            "Keep learning about sustainability to earn more points!",
        ));
    send_embed(&ctx, embed).await
}

/// View the sustainability leaderboard.
#[poise::command(prefix_command, aliases("lb", "top"))]
pub async fn leaderboard(ctx: Context<'_>) -> Result<(), Error> {
    typing(&ctx).await;
    let db = &ctx.data().db;
    let top = UserService::get_leaderboard(db, 10).await?;
    if top.is_empty() {
        return say(&ctx, "❌ No users found on leaderboard.").await;
    }

    let medals = ["🥇", "🥈", "🥉"];
    let mut embed = serenity::CreateEmbed::new()
        .title("🏆 Sustainability Champions Leaderboard")
        .description("Top users making a difference:")
        .colour(colour::gold());

    for (i, u) in top.iter().enumerate() {
        let medal = if i < 3 {
            medals[i].to_string()
        } else {
            format!("**{}.**", i + 1)
        };
        embed = embed.field(
            format!("{} {}", medal, u.username.clone().unwrap_or_default()),
            format!(
                "Level {} • {} points • {}",
                u.level,
                u.sustainability_points,
                u.get_rank()
            ),
            false,
        );
    }

    let me = UserService::get_by_discord_id(db, ctx.author().id.get() as i64).await?;
    if let Some(cur) = me {
        if !top.iter().any(|u| u.id == cur.id) {
            embed = embed.field(
                "\nYour Stats",
                format!(
                    "Level {} • {} points",
                    cur.level, cur.sustainability_points
                ),
                false,
            );
        }
    }
    embed = embed.footer(serenity::CreateEmbedFooter::new(
        "Use !profile to view your complete profile",
    ));
    send_embed(&ctx, embed).await
}

/// Set your sustainability preferences.
#[poise::command(prefix_command, aliases("pref"))]
pub async fn setpreference(
    ctx: Context<'_>,
    #[description = "materials | budget | priority"] preference_type: String,
    #[rest]
    #[description = "Value"]
    value: String,
) -> Result<(), Error> {
    let db = &ctx.data().db;
    let mut user =
        UserService::get_or_create(db, ctx.author().id.get() as i64, &ctx.author().name).await?;

    match preference_type.to_lowercase().as_str() {
        "materials" | "material" => {
            user.preferred_materials = Some(value.clone());
            say(&ctx, format!("✅ Preferred materials set to: {value}")).await?;
        }
        "budget" => {
            if !matches!(value.as_str(), "$" | "$$" | "$$$" | "$$$$") {
                return say(&ctx, "❌ Budget must be: $, $$, $$$, or $$$$").await;
            }
            user.budget_range = Some(value.clone());
            say(&ctx, format!("✅ Budget range set to: {value}")).await?;
        }
        "priority" => {
            let valid = ["environmental", "social", "animal_welfare", "all"];
            let lower = value.to_lowercase();
            if !valid.contains(&lower.as_str()) {
                return say(
                    &ctx,
                    format!("❌ Priority must be one of: {}", valid.join(", ")),
                )
                .await;
            }
            user.sustainability_priority = Some(lower);
            say(&ctx, format!("✅ Sustainability priority set to: {value}")).await?;
        }
        _ => {
            return say(
                &ctx,
                "❌ Invalid preference type. Use: materials, budget, or priority",
            )
            .await
        }
    }

    UserService::update(db, &user).await?;
    Ok(())
}
